### PR TITLE
RHDEVDOCS-4426 - Loki Retention

### PR DESCRIPTION
**Pull request by libander**
Version(s): 4.11+

Issue: [RHDEVDOCS-4426](https://issues.redhat.com//browse/RHDEVDOCS-4426)

Link to docs preview:

QE review:
- [x] QE has approved this change.

Additional information:

### CODE REVIEW

File: modules/logging-loki-retention.adoc [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// Module included in the following assemblies:
+//
+:_content-type: REFERENCE
+[id="logging-loki-retention_{context}"]
+
+.Example LokiStack CR 
+
+[source,yaml]
+----
+apiVersion: loki.grafana.com/v1
```
Comment (kabirbhartiRH) [inline modules/logging-loki-retention.adoc:10]: This look like the correct LokiStack for 'per-tenant stream based retention'. We can also define global stream based retention' for LokiStack.

For e.g.
 ```
apiVersion: loki.grafana.com/v1
 kind: LokiStack
 metadata:
   name: lokistack-sample
   namespace: openshift-logging
 spec:
   limits:
     global:
       retention:
         days: 2
         streams:
         - days: 4
           priority: 1
           selector: '{kubernetes_namespace_name=~"test.+"}'
         - days: 1
           priority: 1
           selector: '{log_type="application"}'
         - days: 15
           priority: 1
           selector: '{log_type="audit"}'
   managementState: Managed
   replicationFactor: 1
   size: 1x.small
   storage:
    schemas:
    - effectiveDate: "2020-10-11"
       version: v11
    secret:
      name: gcs-secret
      type: gcs
  storageClassName: standard
  tenants:
    mode: openshift-logging
```

File: modules/logging-loki-retention.adoc [hunk 2]
```diff
@@ -0,0 +1,76 @@
+// Module included in the following assemblies:
+//
+:_content-type: PROCEDURE
+[id="logging-loki-retention_{context}"]
+= Enabling stream based retention with Loki
```
Comment (rolfedh) [inline modules/logging-loki-retention.adoc:5]:
```suggestion
= Enabling stream-based retention with Loki
```

File: modules/logging-loki-retention.adoc [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+With Logging version 5.6 and higher, you can configure retention policies based on stream globally, per tenant, or both. If you configure both, tenant rules apply before global rules.
```
Comment (rolfedh) [inline modules/logging-loki-retention.adoc:6]: "based on stream globally" sounds a bit odd.

File: modules/logging-loki-retention.adoc [hunk 4]
```diff
@@ -0,0 +1,76 @@
+// Module included in the following assemblies:
+//
+:_content-type: PROCEDURE
+[id="logging-loki-retention_{context}"]
+= Enabling stream based retention with Loki
+With Logging version 5.6 and higher, you can configure retention policies based on stream globally, per tenant, or both. If you configure both, tenant rules apply before global rules. 
+
+. To enable stream based retention,  create or edit the `LokiStack` custom resource:
```
Comment (rolfedh) [inline modules/logging-loki-retention.adoc:8]:
```suggestion
. To enable stream-based retention, create or edit the `LokiStack` custom resource (CR):
```

**Review by rolfedh — APPROVED** [review]
I've offered some suggestions. Pre-approving to expedite.

File: modules/logging-loki-retention.adoc [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        tenantb:
+          retentionLimits:
+            periodDays:
+            stream:
+              - selector:
+                  name: container
+                  type: equal
+                  value: nginx
+                priority: 1
+                periodDays: 1
```
Comment (QiaolingTang) [inline modules/logging-loki-retention.adoc:66]: This is not a valid LokiStack CR, many fields are not in the lokistack CRD: https://github.com/openshift/loki/blob/release-5.6/operator/bundle/manifests/loki.grafana.com_lokistacks.yaml.

Some examples:
1. Per-Tenant stream-based retention
```
apiVersion: loki.grafana.com/v1
kind: LokiStack
metadata:
  name: lokistack-sample
  namespace: openshift-logging
spec:
  limits:
    global:
      retention:
        days: 20
    tenants:
      application:
        retention:
          days: 1
          streams:
            - days: 4
              selector: '{kubernetes_namespace_name=~"test.+"}'
      infrastructure:
        retention:
          days: 5
          streams:
            - days: 1
              selector: '{kubernetes_namespace_name=~"openshift-cluster.+"}'
  managementState: Managed
  replicationFactor: 1
  size: 1x.extra-small
  storage:
    schemas:
    - effectiveDate: "2020-10-11"
      version: v11
    secret:
      name: gcs-secret
      type: gcs
  storageClassName: standard
  tenants:
    mode: openshift-logging 
```
2. Global stream-based retention
```
 apiVersion: loki.grafana.com/v1
 kind: LokiStack
 metadata:
   name: lokistack-sample
   namespace: openshift-logging
 spec:
   limits:
     global:
       retention:
         days: 20
         streams:
         - days: 4
           priority: 1
           selector: '{kubernetes_namespace_name=~"test.+"}'
         - days: 1
           priority: 1
           selector: '{log_type="infrastructure"}'
   managementState: Managed
```

File: modules/logging-loki-retention.adoc [hunk 6]
```diff
@@ -0,0 +1,76 @@
+// Module included in the following assemblies:
+//
+:_content-type: PROCEDURE
+[id="logging-loki-retention_{context}"]
+= Enabling stream-based retention with Loki
+With Logging version 5.6 and higher, you can configure retention policies based on stream. Rules for these may be set globally, per tenant, or both. If you configure both, tenant rules apply before global rules.
```
Comment (periklis) [inline modules/logging-loki-retention.adoc:6]:
```suggestion
With Logging version 5.6 and higher, you can configure retention policies based on log streams. Rules for these may be set globally, per tenant, or both. If you configure both, tenant rules apply before global rules.
```

File: modules/logging-loki-retention.adoc [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+              - selector:
+                  name: container
+                  type: equal
+                  value: nginx
+                priority: 1
+                periodDays: 1
+----
+<1> Retention is enabled in the cluster when this block is added to the CR.
+<2> Time after which the compactor will delete marked chunks.
+<3> Sets retention policy for all logs.
```
Comment (periklis) [inline modules/logging-loki-retention.adoc:70]:
```suggestion
<3> Sets retention policy for all log streams.
```

We should highlight in a note below that this field is not to be used as the global retention policy of 30days that OpenShift Logging includes as a maximum supported policy. The global 30 days policy across all stored logs has to be defined on the Object Storage provider side for the LokiStack bucket. The key wording here is:
- OpenShift Logging 30 days retention -> Affects **stored logs** on object storage -> Configuration on the object storage provider only
- The field mentioned in `<3>` is a global **log stream** retention policy and applying it is at best effort by the compactor. In addition if someone mixes tenant with global retention policies the global policy is overridden anyway.

**Review by periklis — CHANGES_REQUESTED** [review]

File: modules/logging-loki-retention.adoc [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+.Example global stream-based retention
+[source,yaml]
+----
+apiVersion: loki.grafana.com/v1
+kind: LokiStack
+metadata:
+  name: lokistack-sample
+  namespace: openshift-logging
+spec:
+    limits:
```
Comment (QiaolingTang) [inline modules/logging-loki-retention.adoc:23]: Too many blanks before `limits`, here should be:
```
spec:
  limits:
    global:
```

**Review by periklis — COMMENTED** [review]
Besides what @QiaolingTang marked, looks good to me. I have not tested or validated the YAML TBH. The remarks I asked for are in place. Thank you so much :tada:

**Comment by QiaolingTang** [comment]
The changes LGTM.

**Pull request merged**